File: Ad_Hoc/0_1_Knapsack/Knapsack.java
import java.util.Scanner;

public class Knapsack
{
    public static int get_max_value( int[] values, int[] weights, int capacity )
    {
        int[][] knapsack_values = new int[values.length + 1][capacity + 1];

        for( int i = 1; i <= values.length; i++ )
            for( int j = 0; j <= capacity; j++ )
            {
                if( weights[i - 1] <= j )
                    knapsack_values[i][j] = Math.max( knapsack_values[i - 1][j], values[i - 1] + knapsack_values[i - 1][j - weights[i - 1]] );
                else
                    knapsack_values[i][j] = knapsack_values[i - 1][j];
            }

        return knapsack_values[values.length][capacity];
    }
        
    public static void main( String[] args )
    {
        Scanner in = new Scanner( System.in );
        int capacity = in.nextInt(), num_objects = in.nextInt();

        int[] values = new int[num_objects];
        int[] weights = new int[num_objects];

        int tmp;
        for( int i = 0; i < num_objects; i++ )
        {
            tmp = in.nextInt();
            values[i] = tmp;
        }

        for( int i = 0; i < num_objects; i++ )
        {
            tmp = in.nextInt();
            weights[i] = tmp;
        }

        int max_value = get_max_value( values, weights, capacity );
        System.out.println( "The maximum value you can take from the knapsack is " + max_value );
    }
};

File: knapsack.cpp
#include <iostream>
#include <vector>
using namespace std;

int knapsack( const int& capacity, vector <int> values, vector <int> weights )
{
	vector <int> max_value_temp( capacity + 1, 0 );
	vector < vector <int> > max_value( values.size() + 1, max_value_temp );

	for( int i = 1; i <= values.size(); i++ )
		for( int j = 1; j <= capacity; j++ )
		{
			if( weights[i - 1] <= j )
				max_value[i][j] = max( max_value[i - 1][j], values[i - 1] + max_value[i - 1][j - weights[i - 1]] );
			else
				max_value[i][j] = max_value[i - 1][j];
		}

	return max_value[values.size()][capacity];
}

int main()
{
	int num_items, capacity;
	cin >> num_items >> capacity;

	vector <int> values( num_items );
	for( int item_number = 0; item_number < num_items; item_number++ )
		cin >> values[item_number];
	
	vector <int> weights( num_items );
	for( int item_number = 0; item_number < num_items; item_number++ )
		cin >> weights[item_number];

	cout << "The maximum value that can fit within the knapsack's capacity is " << knapsack( capacity, values, weights ) << endl;

	return 0;
}
